### PR TITLE
Add multiple regions analysis / 5R / SMURF / q2-sidle

### PR DESCRIPTION
**Pull request by d4straub**
Addresses https://github.com/nf-core/ampliseq/issues/701

Status:
- [x] produce simulated small test data
- [x] move primer info into meta map, i.e. sample-wise
- [x] process primers sample-wise
- [x] read multi region info via --input_multiregion
- [x] process without error `--input samplesheet_multiregion.tsv --input_multiregion regions_multiregion.tsv --skip_taxonomy`
- [x] retain all current function
- [x] produce per-region ASV table
- [x] add reconstruction subworkflow via pipesidle
- [x] plugin sidle output to downstream analysis
- [x] add sidle reference taxonomy entries & custom input
- [x] process `--sidle_ref_tax_custom <file,file,file> --sidle_ref_tree_custom <file>`
- [x] process `--sidle_ref_taxonomy silva`
- [x] process `--sidle_ref_taxonomy greengenes`
- [x] add proper test
- [x] ~~q2-sidle to conda~~ not wanted by q2-sidle developer
- [x] update docs
- [x] make sure all tests pass [except linting, template update prevents that!]
- [x] organize output
- [x] organize params in nextflow_schema.json
- [x] add params check for incompatible settings in `lib/WorkflowAmpliseq.groovy`

**Potential problem:**
- All processes using Sidle use `container 'docker.io/d4straub/pipesidle:0.1.0-beta'`, a container I build with https://github.com/d4straub/pipesidle/blob/main/Dockerfile and hosted in my docker hub account. The author of the pypi package doesnt want the software in conda. So thats the way I went for now. Any suggestions welcome.
- Failing linting is due to template update 2.13 (that restructures pipeline functions in folder lib, therefore postponed for now).

Ideas postponed:
- optional per-region read trimming definitions (trunclenf & trunlenr)
- In case there are multiple regions, in `overall_summary.tsv` & `*.stats.tsv` remove from sample the region info and aggregate to obtain numbers per-sample instead of currently per-sample-per-region.
- Pipeline summary report isnt compatible to multi-region analysis, this could be improved in the future
- use appropriate taxlevels (`conf/ref_databases.conf` see `taxlevels`) & output tab separated taxonomies

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/ampliseq/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, also make a PR on the nf-core/ampliseq _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by d4straub** [comment]
Failing linting is due to template update 2.13 (that restructures pipeline functions in folder lib, therefore postponed for now).

**Comment by d4straub** [comment]
The implemented test_multiregion seems to run quite long, here are the processes with longest walltime when I run it on my laptop:

process | min | sec
-- | -- | --
NFCORE_AMPLISEQ:AMPLISEQ:QIIME2_DIVERSITY:QIIME2_ALPHARAREFACTION | 16 | 14
NFCORE_AMPLISEQ:AMPLISEQ:SIDLE_WF:SIDLE_TREERECON | 13 | 21
NFCORE_AMPLISEQ:AMPLISEQ:SIDLE_WF:SIDLE_DBEXTRACT | 5 | 37
NFCORE_AMPLISEQ:AMPLISEQ:SIDLE_WF:SIDLE_DBFILT | 5 | 29

SIDLE_DBEXTRACT is repeated for each region, i.e. in this example 5 times.

Possible changes to reduce runtime:
- 16min: `--skip_alpha_rarefaction`, no disadvantage
- 13min+: SIDLE_TREERECON, currently caanot be omitted; disadvantage if omitted: phylogenetic tree isnt constructed, this will lead omission of downstream processes involving alpha & beta diversity
- 10min+: SIDLE_DBEXTRACT, using less regions, i.e. less repetitions of this process, potentially downstream processes faster
- 10min+: SIDLE_DBEXTRACT, using smaller database, i.e. greengenes 85% or such, disadvantage: minimal!

File: bin/taxref_reformat_sidle.sh [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    # remove uncompressed folder
+    rm -r SILVA_128_QIIME_release
+else
+    echo "No expected directory detected"
+fi
+
+
+
+
+
```
Comment (erikrikarddaniel) [inline bin/taxref_reformat_sidle.sh:28]: Should these be there?

File: modules/local/dada2_err.nf [hunk 2]
```diff
@@ -34,32 +35,32 @@ process DADA2_ERR {
         fnFs <- sort(list.files(".", pattern = "_1.filt.fastq.gz", full.names = TRUE))
         fnRs <- sort(list.files(".", pattern = "_2.filt.fastq.gz", full.names = TRUE))
 
-        sink(file = "${meta.run}.err.log")
+        sink(file = "${prefix}.err.log")
```
Comment (erikrikarddaniel) [inline modules/local/dada2_err.nf:38]: I'm not sure I understand how this works. I would've guessed that you need also `${meta.run}` or is the run set in the config as `task.ext.prefix` now?

**Review by erikrikarddaniel — APPROVED** [review]
Two tiny comments. Otherwise :+1:

File: bin/taxref_reformat_sidle.sh (hunk 1 above)
Comment (d4straub) [reply bin/taxref_reformat_sidle.sh:28]: No, will remove that!

File: modules/local/dada2_err.nf (hunk 2 above)
Comment (d4straub) [reply modules/local/dada2_err.nf:38]: Exactely, its now defined in `conf/modules.config` as `ext.prefix = { meta.region ? "region-${meta.region}_run-${meta.run}" : "${meta.run}" }`!

**Pull request merged**